sbnanalysis/ana/SBNOsc/NumuSelection.cxx

```diff
@@ -37,31 +145,228 @@ bool NumuSelection::ProcessEvent(const gallery::Event& ev, std::vector<Event::In
               << "(" << fNuCount << " neutrinos selected)"
               << std::endl;
   }
+
+  // clean up containers
   fEventCounter++;
+  _interactionInfo->clear();
 
-  // Grab a data product from the event
+  // Get truth
   auto const& mctruths = \
-    *ev.getValidHandle<std::vector<simb::MCTruth> >(fTruthTag);
+    *ev.getValidHandle<std::vector<simb::MCTruth> >(_config.truthTag);
 
   // Iterate through the neutrinos
+  bool selected = false;
   for (size_t i=0; i<mctruths.size(); i++) {
-    Event::Interaction interaction;
     auto const& mctruth = mctruths.at(i);
+    // get the neutrino
     const simb::MCNeutrino& nu = mctruth.GetNeutrino();
 
-    if (nu.CCNC() == simb::kCC && nu.Mode() == 0 && nu.Nu().PdgCode() == 14) {
-      Event::Interaction interaction = TruthReco(mctruth);
+    // build the interaction
+    Event::Interaction interaction = TruthReco(mctruth);
+
+    // Get selection-specific info
+    //
+    // Start with the interaction stuff
+    auto intInfo = interactionInfo(ev, mctruth);
+
+    // run selection
+    std::vector<bool> selection = Select(ev, mctruth, i, intInfo);
+
+    // pass iff pass each cut
+    bool pass_selection = std::find(selection.begin(), selection.end(), false) == selection.end();
+    if (pass_selection) { 
+      // store interaction in reco tree
       reco.push_back(interaction);
+      // store local info
+      _interactionInfo->push_back(intInfo);
+
+      fNuCount++;
+      selected = true;
+    }
+    
+    // fill histos
+    unsigned select_i = 0; 
+    for (bool pass: selection) {
+      if (pass) {
+        _root_histos[select_i].h_numu_trueE->Fill(interaction.neutrino.energy);
+        _root_histos[select_i].h_numu_ccqe->Fill(ECCQE(interaction.lepton.momentum, interaction.lepton.energy));
+        _root_histos[select_i].h_numu_visibleE->Fill(intInfo.visible_energy);
+        _root_histos[select_i].h_numu_true_v_visibleE->Fill(intInfo.visible_energy - interaction.neutrino.energy);
+        _root_histos[select_i].h_numu_l_length->Fill(intInfo.l_length);
+        _root_histos[select_i].h_numu_contained_L->Fill(intInfo.l_contained_length);
+        _root_histos[select_i].h_numu_l_is_contained->Fill(intInfo.l_is_contained);
+        _root_histos[select_i].h_numu_Vxy->Fill(nu.Nu().Vx(), nu.Nu().Vy());
+        _root_histos[select_i].h_numu_Vxz->Fill(nu.Nu().Vx(), nu.Nu().Vz());
+        _root_histos[select_i].h_numu_Vyz->Fill(nu.Nu().Vy(), nu.Nu().Vz());
+      }
+      select_i ++;
```

Comment:
It seems like it would be more clear to use an index-based loop here.

Reply:
Fixed.